Don't load global state when only restoring indices

tlrx opened the pull request:
Restoring a snapshot (or getting the status of finished snapshots) currently always load the global state metadata file from the repository even if it not required. This slows down the restore process (or listing statuses process)  and can also be an issue if the global state cannot be deserialized (because it has unknown customs for example).

This commit splits the `Repository.getSnapshotMetadata()` method into two distincts methods: `getGlobalMetadata()` and `getIndexMetadata()` that are now called on purpose. It also adds a test to control which/when metadata files are accessed. I'm not really happy of the test but I didn't come up with a better one. It also renames few variables and reorganizes the order of some instructions.

Closes #28934

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:
```diff
@@ -182,17 +184,36 @@ public void restoreSnapshot(final RestoreRequest request, final ActionListener<R
+            }
+
+            final List<IndexId> indexIdsInSnapshot = repositoryData.resolveIndices(indicesInSnapshot);
+            if (indexIdsInSnapshot != null) {
```
should never be null?

tlrx replied:
Right, it should never be null.

ywelsch commented on server/src/test/java/org/elasticsearch/snapshots/MetadataLoadingDuringSnapshotRestoreIT.java:
```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    private void assertIndexMetadataLoads(final String snapshot, final String index, final Integer times) {
```
use an int instead of an integer and use 0 to denote that indexmetadata was never loaded

tlrx replied:
I wonder why I did this. Thanks

ywelsch commented on server/src/test/java/org/elasticsearch/snapshots/MetadataLoadingDuringSnapshotRestoreIT.java:
```diff
@@ -0,0 +1,209 @@
+
+/**
+ * This class tests that snapshot's global metadata and index metadata are loaded from the
+ * {@link Repository} on purpose
```
maybe instead: tests whether global and index metadata are only loaded from the repository when needed.

tlrx commented:
Thanks @ywelsch for the review, I updated the code with your feedback.

> What about deleting snapshots not requiring to load global state (as mentioned in #28934)? Will you do that as a follow-up?

Yes, I'd like to look at this in detail in another pull request.

> Regarding tests, an alternative (or addition to the existing) test is to corrupt the global metadata and then check that you can still restore indices (see testListCorruptedSnapshot for inspiration). Similarly, corrupt one index and check that you can still restore another index in same snapshot.

That's a great suggestion, I added this two tests.

tlrx commented on server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -2590,12 +2592,152 @@ public void testListCorruptedSnapshot() throws Exception {
+    public void testRestoreSnapshotWithCorruptedIndexMetadata() throws Exception {
+        final Client client = client();
+        final Path repo = randomRepoPath();
+        final int nbIndices = randomIntBetween(2, 7);
```
Sure

tlrx commented:
Thanks @ywelsch and @imotov

tlrx commented:
This change has been backported to 6.x in 9a065af4da0c176e8a63b6134396b4057dd17729
